File: crates/zipp-vm/tests/json_ascii_unchecked.js
// JSON ASCII-unchecked / integer-token oracle: `JSON.parse` and
// `JSON.stringify` must produce the same bytes whether a parser-proved ASCII
// view, a direct integer read, or the general `from_utf8` + `parse::<f64>`
// path served them. Every case here is one where the shortcut and the
// general path could plausibly disagree: non-ASCII and escaped names, lone
// surrogates, controls, `-0`, 15/16/17-digit integers, fractions, exponents,
// ropes, toJSON and revivers.
var out = [];
function rec(label, v) { out.push(label + ":" + v); }
function rt(s) { return JSON.stringify(JSON.parse(s)); }

// A. plain ASCII names and values round-trip byte-for-byte
var a = '{"alpha":1,"beta_2":"x-17","g":[true,false,null],"":""}';
rec("A", rt(a) + "|" + (rt(a) === a));

// B. non-ASCII names and values (2-, 3- and 4-byte UTF-8)
var b = '{"é":"café","中文":"😀","k":"aéb"}';
var bo = JSON.parse(b);
rec("B", rt(b) + "|" + Object.keys(bo).join(",") + "|" + bo["中文"].length);

// C. escapes in names and values: NUL, controls, quotes, backslash, slash
var c = '{"n\\u0000l":"\\u0000\\n\\t\\"\\\\\\/","\\u0001":"\\u001f\\b\\f\\r"}';
var co = JSON.parse(c);
rec("C", rt(c) + "|" + Object.keys(co).map(function (k) { return k.length; }).join(","));

// D. lone surrogates: escaped in the source, raw in a JS string. (A lone
// surrogate in a member NAME is a known pre-existing U+FFFD fold — B233's
// side-finding — the same with every latch, so it is not pinned here.)
var d = JSON.parse('{"s":"\\ud800x\\udc00","k":"\\udbff"}');
rec("D", JSON.stringify(d) + "|" + d.s.length + "|" + d.s.charCodeAt(0) + "|" +
  JSON.stringify("a\ud800b") + "|" + JSON.stringify({ q: "\udc00\ud83d" }));

// E. numbers: the integer fast lane's exact boundary and everything past it
var e = JSON.parse('[0,-0,7,-7,123456789012345,-999999999999999,1000000000000000,' +
  '9007199254740993,12345678901234567890123,1.5,-2.5,0.1,1e5,1E-2,1e21,1e-7,-1.25e+3,0.000001,1e400,-1e400]');
rec("E", JSON.stringify(e) + "|" + (1 / e[1]) + "|" + Object.is(e[1], -0) + "|" +
  e.map(function (n) { return typeof n; }).join(""));

// F. toJSON and a reviver see the same values on either path
var f = { a: 1, b: { toJSON: function () { return "tj-" + this.x; }, x: 5 }, d: new Date(0), s: "q" };
rec("F", JSON.stringify(f) + "|" + JSON.stringify(JSON.parse('{"a":1,"b":[2,3],"c":"z"}',
  function (k, v) { return typeof v === "number" ? v * 10 : k === "c" ? undefined : v; })));

// G. stringify strings: ASCII with escapes, a rope, DEL and U+2028/2029, long ASCII
var g1 = "he said \"hi\"\\ " + String.fromCharCode(7, 27) + " end";
var rope = ""; for (var i = 0; i < 40; i++) rope += "ab" + i;
var g3 = "x" + String.fromCharCode(0x7f) + "y" + String.fromCharCode(0x2028) + "z" + String.fromCharCode(0x2029) + "w"; // DEL and the line terminators are NOT escaped
var longA = ""; for (var j = 0; j < 300; j++) longA += String.fromCharCode(32 + (j % 95));
var gs = JSON.stringify([g1, rope, g3, longA]);
function codes(s) { var r = []; for (var ci = 0; ci < s.length; ci++) r.push(s.charCodeAt(ci)); return r.join("."); }
rec("G", gs.length + "|" + codes(JSON.stringify(g1)) + "|" + codes(JSON.stringify(g3)) + "|" +
  (JSON.parse(gs)[3] === longA) + "|" + JSON.parse(gs)[1].length);

// H. name ordering: canonical indices first, then textual; empty name and value
var h = JSON.parse('{"b":1,"10":2,"2":3,"":4,"-1":5,"01":6,"a":""}');
rec("H", Object.keys(h).join(",") + "|" + JSON.stringify(h));

// I. a bench-shaped tree, stringify -> parse -> stringify fixed point
function mulberry32(seed) { var s = seed | 0; return function () {
  s = (s + 0x6D2B79F5) | 0; var t = Math.imul(s ^ (s >>> 15), 1 | s);
  t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t; return ((t ^ (t >>> 14)) >>> 0) / 4294967296; }; }
var rnd = mulberry32(7); function ri(n) { return (rnd() * n) | 0; }
function build(depth) {
  if (depth >= 4) { var t = ri(6);
    return t === 0 ? ri(1000000) : t === 1 ? ri(2000) + 0.5 : t === 2 ? "w" + ri(1000) + "é" : t === 3 ? true : t === 4 ? -ri(50) : null; }
  if (ri(2) === 0) { var arr = []; for (var k = 0; k < 3; k++) arr.push(build(depth + 1)); return arr; }
  var o = {}; for (var m = 0; m < 3; m++) o["k" + ri(20) + "_" + m] = build(depth + 1); return o;
}
var tree = []; for (var r = 0; r < 12; r++) tree.push(build(0));
var s1 = JSON.stringify(tree), s2 = JSON.stringify(JSON.parse(s1));
var hh = 0x811c9dc5; for (var q = 0; q < s1.length; q++) hh = Math.imul(hh ^ s1.charCodeAt(q), 16777619);
rec("I", s1.length + "|" + (s1 === s2) + "|" + (hh >>> 0));

// J. number-grammar errors and boundaries still throw / parse as before
function tryParse(s) { try { return JSON.stringify(JSON.parse(s)); } catch (e) { return e.name; } }
rec("J", ["-", "-0", "01", "1.", ".5", "1e", "1e+", "0e0", "-0.0", "1E+2", "[1,]", '"\\ud800"', '"a"']
  .map(tryParse).join(","));

console.log(out.join("\n"));

File: crates/zipp-vm/tests/json_ascii_unchecked.rs
//! `JSON.parse`/`JSON.stringify` build `&str` views of bytes the scanner has
//! already proved ASCII without a second `core::str::from_utf8` pass — an
//! escape-free member name whose scan saw no byte >= 0x80, a number token
//! (ASCII by grammar), and a flat heap string whose `JsStr::ascii` flag is set
//! — and a plain integer token of at most 15 digits is read directly instead
//! of through `str::parse::<f64>`.
//!
//! Both shortcuts must be INVISIBLE: the same bytes out of `stringify`, the
//! same values out of `parse`, the same errors. The JS file attacks exactly
//! the places the shortcut and the general path could part ways — non-ASCII
//! and escaped names, lone surrogates, controls, `-0`, the 15/16/17-digit
//! boundary, fractions and exponents, ropes, `toJSON`, revivers, grammar
//! errors — and every expectation is node-oracled (v24.12.0).
//!
//! `ZIPP_NO_JSON_ASCII_UNCHECKED=1` and `ZIPP_NO_JSON_INT_FAST=1` restore the
//! old paths; the child runs below check the same file under each latch and
//! under every execution mode, so a divergence is attributable.

const SRC: &str = include_str!("json_ascii_unchecked.js");

const EXPECTED: &str = concat!(
    "A:{\"alpha\":1,\"beta_2\":\"x-17\",\"g\":[true,false,null],\"\":\"\"}|true\n",
    "B:{\"\u{e9}\":\"caf\u{e9}\",\"\u{4e2d}\u{6587}\":\"\u{1f600}\",\"k\":\"a\u{e9}b\"}|\u{e9},\u{4e2d}\u{6587},k|2\n",
    "C:{\"n\\u0000l\":\"\\u0000\\n\\t\\\"\\\\/\",\"\\u0001\":\"\\u001f\\b\\f\\r\"}|3,1\n",
    "D:{\"s\":\"\\ud800x\\udc00\",\"k\":\"\\udbff\"}|3|55296|\"a\\ud800b\"|{\"q\":\"\\udc00\\ud83d\"}\n",
    "E:[0,0,7,-7,123456789012345,-999999999999999,1000000000000000,9007199254740992,",
    "1.2345678901234568e+22,1.5,-2.5,0.1,100000,0.01,1e+21,1e-7,-1250,0.000001,null,null]",
    "|-Infinity|true|numbernumbernumbernumbernumbernumbernumbernumbernumbernumber",
    "numbernumbernumbernumbernumbernumbernumbernumbernumbernumber\n",
    "F:{\"a\":1,\"b\":\"tj-5\",\"d\":\"1970-01-01T00:00:00.000Z\",\"s\":\"q\"}|{\"a\":10,\"b\":[20,30]}\n",
    "G:510|34.104.101.32.115.97.105.100.32.92.34.104.105.92.34.92.92.32.92.117.48.48.48.55.92.117.",
    "48.48.49.98.32.101.110.100.34|34.120.127.121.8232.122.8233.119.34|true|150\n",
    "H:2,10,b,,-1,01,a|{\"2\":3,\"10\":2,\"b\":1,\"\":4,\"-1\":5,\"01\":6,\"a\":\"\"}\n",
    "I:11943|true|432579531\n",
    "J:SyntaxError,0,SyntaxError,SyntaxError,SyntaxError,SyntaxError,SyntaxError,0,0,100,",
    "SyntaxError,\"\\ud800\",SyntaxError",
);

#[test]
fn json_round_trips_match_node() {
    let out = zipp_vm::run(SRC).expect("source compiles");
    assert!(out.error.is_none(), "{:?}", out.error);
    assert_eq!(out.output, vec![EXPECTED.to_string()]);
}

/// The same program under each latch and each execution mode, in a child
/// process (the latches are read once per process).
#[test]
fn json_round_trips_match_in_every_mode_and_with_the_latches_off() {
    if std::env::var_os("ZIPP_JSON_ASCII_CHILD").is_some() {
        return;
    }
    let exe = std::env::current_exe().expect("test binary path");
    const LATCHES: [&str; 8] = [
        "ZIPP_NO_JSON_ASCII_UNCHECKED",
        "ZIPP_NO_JSON_INT_FAST",
        "ZIPP_NO_JSON_PLAIN_KEY",
        "ZIPP_NO_JSON_QUOTE_BULK",
        "ZIPP_NO_JSON_LEAF_FAST",
        "ZIPP_NOJIT",
        "ZIPP_JIT_THRESHOLD",
        "ZIPP_NO_NURSERY",
    ];
    for (mode, env) in [
        ("default", None),
        ("no-ascii-unchecked", Some(("ZIPP_NO_JSON_ASCII_UNCHECKED", "1"))),
        ("no-int-fast", Some(("ZIPP_NO_JSON_INT_FAST", "1"))),
        ("no-plain-key", Some(("ZIPP_NO_JSON_PLAIN_KEY", "1"))),
        ("no-quote-bulk", Some(("ZIPP_NO_JSON_QUOTE_BULK", "1"))),
        ("no-leaf-fast", Some(("ZIPP_NO_JSON_LEAF_FAST", "1"))),
        ("interpreter", Some(("ZIPP_NOJIT", "1"))),
        ("forced-jit", Some(("ZIPP_JIT_THRESHOLD", "1"))),
        ("no-nursery", Some(("ZIPP_NO_NURSERY", "1"))),
    ] {
        let mut cmd = std::process::Command::new(&exe);
        cmd.args(["--exact", "json_round_trips_match_node", "--nocapture"])
            .env("ZIPP_JSON_ASCII_CHILD", "1");
        for l in LATCHES {
            cmd.env_remove(l);
        }
        if let Some((key, value)) = env {
            cmd.env(key, value);
        }
        let out = cmd.output().expect("spawn mode child");
        assert!(
            out.status.success(),
            "{mode} failed:\n--- stdout ---\n{}\n--- stderr ---\n{}",
            String::from_utf8_lossy(&out.stdout),
            String::from_utf8_lossy(&out.stderr)
        );
    }
}
